### PR TITLE
Feature/POPUG-HW-03

### PR DESCRIPTION
**Pull request by Ballr0g**
Implemented employee-billing service:
- CUD TaskCreated event listener for task replication
- Business event listeners:
    - TaskCreatedEvent - initiates billing of a specific user.
    - TaskReassignedEvent - reassigns a task, bills a new user for assignment.
    - TaskCompletedEvent - closes a task, pays user for completion.
- Implemented billing cycles - automatically initiates the first on app start, closes cycles and pays at 12AM UTC daily.

Todo:
- Schema registry.
- Avro schema + versioning.
- Analytics.

Known problems:
- Race condition between CUD and business TaskCreated events - planning to handle it by storing TaskCreated business events while waiting for appropriate CUD events before them.
- No guarantees of exactly-once message handling in listeners - kafka failures results in billing happening more than once. Planning to handle it by adding a unique event id for every event.

### CODE REVIEW

File: employee-billing/src/main/java/org/uber/popug/employee/billing/domain/billing/creation/TaskForBillingAssignment.java [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+import jakarta.annotation.Nonnull;
+
+import java.time.LocalDateTime;
+import java.util.UUID;
+
+// Todo: rename - org.uber.popug.employee.billing.domain.task.creation.TaskAssignmentInfo due to its general nature.
+public record TaskForBillingAssignment(
+        @Nonnull UUID publicId,
+        @Nonnull UUID assigneeId,
```
Comment (davydovanton) [inline employee-billing/src/main/java/org/uber/popug/employee/billing/domain/billing/creation/TaskForBillingAssignment.java:11]: тут тоже публичный assignee id должен быть тогда (я про нейминг, не про тип данных)

**Review by davydovanton — COMMENTED** [review]
привет, все посмотрел, названия топиков только поменять, в остальном вопросов нет

File: employee-billing/src/main/java/org/uber/popug/employee/billing/config/KafkaConsumerConfig.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+import org.uber.popug.employee.billing.kafka.TasksBusinessWorkflowJsonDeserializer;
+import org.uber.popug.employee.billing.kafka.TasksCUDJsonDeserializer;
+
+import java.util.List;
+import java.util.Map;
+
+@Configuration(proxyBeanMethods = false)
+public class KafkaConsumerConfig {
+
+    @Value("${kafka.listener.task-lifecycle-stream.bootstrap-servers}")
```
[the remark this answers is not in the document]
Comment (Ballr0g) [reply employee-billing/src/main/java/org/uber/popug/employee/billing/config/KafkaConsumerConfig.java:21]: Да, поправил - теперь использую более лаконичные tasks.stream и tasks-workflow.

File: employee-billing/src/main/java/org/uber/popug/employee/billing/domain/billing/creation/TaskForBillingAssignment.java (hunk 1 above)
Comment (Ballr0g) [reply employee-billing/src/main/java/org/uber/popug/employee/billing/domain/billing/creation/TaskForBillingAssignment.java]: Тут не успел зарефакторить, сейчас переделал в едином стиле по аналогии с TaskReassignmentInfo и TaskCompletionInfo. ПО неймингу теперь - taskExtPublicId, assigneeExtPublicId.

**Pull request merged**